core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -831,8 +839,10 @@ private[spark] class SparkSubmit extends Logging {
       }
     }
 
-    if (isKubernetesCluster) {
-      childMainClass = KUBERNETES_CLUSTER_SUBMIT_CLASS
+    if (isKubernetesCluster | isExternalCluster) {
+      childMainClass =
+        if (isKubernetesCluster) KUBERNETES_CLUSTER_SUBMIT_CLASS
+        else sparkConf.get(EXTERNAL_CLUSTER_SUBMIT_CLASS).get
```

Comment:
The existence of this config value should be checked earlier and a meaningful error message should be given if not set.

Reply:
Same as for your remark for the cluster manager, we can do this check only after the JARs are loaded, as those classes will be available then.